zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1304,7 +1356,11 @@ private void fireDataChangedEvents(final String path, Set<IZkDataListenerEntry>
           public void run() throws Exception {
             if (!pathStatRecord.pathChecked()) {
               // getStat will re-install watcher only when the path exists
-              pathStatRecord.recordPathStat(getStat(path, pathExists), notificationTime);
+              if (!pathExists) {
+                pathStatRecord.recordPathStat(getStat(path, false), notificationTime);
+              } else {
+                pathStatRecord.recordPathStat(getStat(path, true, true), notificationTime);
```

Comment:
Old logic, if pathExists is false, it would not install watch using exist() call; otherwise, use exists() call to install the watch.

New logic, if pathExits is false, it would not install watch using exist() call; otherwise, use getData() call to istall the watch.

New approach would not leak watch if async deletion happens in between.